eth/watchers/pricefeedwatcher.go

```diff
@@ -0,0 +1,176 @@
+package watchers
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/livepeer/go-livepeer/clog"
+	"github.com/livepeer/go-livepeer/eth"
+)
+
+const (
+	priceUpdateMaxRetries     = 5
+	priceUpdateBaseRetryDelay = 30 * time.Second
+	priceUpdatePeriod         = 1 * time.Hour
+)
+
+// PriceFeedWatcher monitors a Chainlink PriceFeed for updated pricing info. It
+// allows fetching the current price as well as listening for updates on the
+// PriceUpdated channel.
+type PriceFeedWatcher struct {
+	ctx            context.Context
+	baseRetryDelay time.Duration
+
+	priceFeed                   eth.PriceFeedEthClient
+	currencyBase, currencyQuote string
+
+	current      eth.PriceData
+	priceUpdated chan eth.PriceData
+}
+
+// NewPriceFeedWatcher creates a new PriceFeedWatcher instance. It will already
+// fetch the current price and start a goroutine to watch for updates.
+func NewPriceFeedWatcher(ctx context.Context, rpcUrl, priceFeedAddr string) (*PriceFeedWatcher, error) {
+	priceFeed, err := eth.NewPriceFeedEthClient(ctx, rpcUrl, priceFeedAddr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create price feed client: %w", err)
+	}
+
+	description, err := priceFeed.Description()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get description: %w", err)
+	}
+
+	currencyFrom, currencyTo, err := parseCurrencies(description)
+	if err != nil {
+		return nil, err
+	}
+
+	w := &PriceFeedWatcher{
+		ctx:            ctx,
+		baseRetryDelay: priceUpdateBaseRetryDelay,
+		priceFeed:      priceFeed,
+		currencyBase:   currencyFrom,
+		currencyQuote:  currencyTo,
+		priceUpdated:   make(chan eth.PriceData, 1),
+	}
+
+	err = w.updatePrice()
+	if err != nil {
+		return nil, fmt.Errorf("failed to update price: %w", err)
+	}
+
+	go func() {
+		ctx, cancel := context.WithCancel(w.ctx)
+		defer cancel()
+		ticker := newTruncatedTicker(ctx, priceUpdatePeriod)
+		w.watch(ctx, ticker)
+	}()
+
+	return w, nil
+}
+
+// Currencies returns the base and quote currencies of the price feed.
+// i.e. base = CurrentPrice() * quote
+func (w *PriceFeedWatcher) Currencies() (base string, quote string) {
+	return w.currencyBase, w.currencyQuote
+}
+
+// Current returns the latest fetched price data.
+func (w *PriceFeedWatcher) Current() eth.PriceData {
+	return w.current
+}
+
+// PriceUpdated returns a channel that will receive the updated price data as
+// soon as it changes.
+func (w *PriceFeedWatcher) PriceUpdated() <-chan eth.PriceData {
+	return w.priceUpdated
+}
+
+func (w *PriceFeedWatcher) updatePrice() error {
+	newPrice, err := w.priceFeed.FetchPriceData()
+	if err != nil {
+		return fmt.Errorf("failed to fetch price data: %w", err)
+	}
+
+	if newPrice.UpdatedAt.After(w.current.UpdatedAt) {
+		w.current = newPrice
+		select {
+		case w.priceUpdated <- newPrice:
+		default:
```

Comment:
This is kind-of tricky, because it means that if nobody is consuming the price then we stop sending it to the channel. So, if nobody is consuming and suddenly starts consuming, then it receives the old price.

I think to solve it we should do one of the following:
1. Send to the channel just the information that the price was updated (but without any value)
2. Use the `Subscribe()` mechanism we use in other watchers (as I commented in some other comment)

Reply:
Yeah good points. I switched to an `event.Feed` and explicit `Subscribe()` now, so we wont' have this automatic dropping of events (OTOH we might block if there's no receivers, but the explicit Subscribe/Unsubscribe is enough protection from that IMO).